firedrake/preconditioners/fdm.py

```diff
@@ -0,0 +1,1036 @@
+from functools import lru_cache, partial
+
+from pyop2 import op2
+from pyop2.sparsity import get_preallocation
+
+from ufl import FiniteElement, VectorElement, TensorElement
+from ufl import FacetNormal, Jacobian, JacobianDeterminant, JacobianInverse
+from ufl import as_tensor, diag_vector, dot, dx, indices, inner, inv
+from ufl.algorithms.ad import expand_derivatives
+
+from firedrake.petsc import PETSc
+from firedrake.preconditioners.base import PCBase
+from firedrake.preconditioners.patch import bcdofs
+from firedrake.preconditioners.pmg import get_permuted_map, tensor_product_space_query
+from firedrake.utils import IntType_c
+from firedrake.dmhooks import get_function_space, get_appctx
+import firedrake.dmhooks as dmhooks
+import firedrake
+import numpy
+
+try:
+    from scipy.linalg import eigh
+
+    def sym_eig(A, B):
+        return eigh(A, B)
+except ImportError:
+    import numpy.linalg as npla
+
+    def sym_eig(A, B):
+        L = npla.cholesky(B)
+        Linv = npla.inv(L)
+        C = numpy.dot(Linv, numpy.dot(A, Linv.T))
+        Z, W = npla.eigh(C)
+        V = numpy.dot(Linv.T, W)
+        return Z, V
+
+
+class FDMPC(PCBase):
+
+    _prefix = "fdm_"
+
+    def initialize(self, pc):
+        A, P = pc.getOperators()
+
+        # Read options
+        prefix = pc.getOptionsPrefix()
+        options_prefix = prefix + self._prefix
+        # opts = PETSc.Options(options_prefix)
+
+        dm = pc.getDM()
+        V = get_function_space(dm)
+        use_tensorproduct, N, ndim, family, _ = tensor_product_space_query(V)
+
+        if not use_tensorproduct or (family <= {"RTCE", "NCE"}):
+            raise ValueError("The element %s is not supported by FDMPC." % (V.ufl_element(), ))
+
+        needs_interior_facet = not (family <= {"Q", "Lagrange"})
+        Nq = 2*N+1  # quadrature degree
```

Comment:
Will we ever want to specify a different quadrature degree?

Reply:
The quadrature degree `Nq` is used for two things: 
1. to construct the stiffness and mass matrices on the reference interval
2. to sample the form coefficients on a quadrature element to project them to piecewise constants

For the first task Nq=2*N+1 gives the exact result.
The user might want to speed up things and use a cheaper projection for task 2. To save memory, the form coefficients should be stored in DQ_0, since they are utimately reduced to piecewise constants.

Reply:
Maybe we could use the quadrature degree from the integral from which each coefficient comes from, or the one from the form compiler parameters to average the coefficients?

Reply:
I'm happy hardcoding it. Can you add a comment in the code with what you've said on this thread?